### PR TITLE
JPEG2000: Support decoding subsampled RGB and YCbCr images

### LINKED ISSUE #3946: Truncated/corrupted jpeg2000 image
Hi, I have a batch of images like this and I want them converted to a numpy array:
https://drive.google.com/open?id=1hx_s1LZV48ID89CFaTzE3vjUTXN1IcGl

The images seem to be jpeg2000 and truncated. Trying to load them without LOAD_TRUNCATED_IMAGES gives me the usual exception:
```python
from PIL import Image
a = Image.open('broken.jpg').convert("RGB")
```
OSError: broken data stream when reading image file

Using `LOAD_TRUNCATED_IMAGES = True`
```python
from PIL import Image, ImageFile
import numpy as np
ImageFile.LOAD_TRUNCATED_IMAGES = True
a = np.asarray(Image.open('broken.jpg').convert("RGB"))
print(np.sum(a))
```
Gives a zero sum of the array, indicating that the image is not properly loaded. I was expecting an array containing the RGB information of the image.

* OS: Ubuntu 16.04
* Python: 3.68
* Pillow: 6.1.0

Thank you!

### LINKED ISSUE #4991: OSError: broken data stream when reading image file JP2
I am reading BPG file on pillow version 7.0. I already try many methods from stackoverflow like update pillow and Truncated LOADED. Nothing works

```python
   ## some function borrowed from
## https://github.com/tensorflow/models/blob/master/compression/image_encoder/msssim.py
"""Python implementation of MS-SSIM.
Usage:
python msssim.py --original_image=original.png --compared_image=distorted.png
"""
import argparse

import numpy as np
from scipy import signal
from scipy.ndimage.filters import convolve
from PIL import Image


parser = argparse.ArgumentParser()
parser.add_argument('--metric', '-m', type=str, default='all', help='metric')
parser.add_argument(
    '--original-image', '-o', type=str, required=True, help='original image')
parser.add_argument(
    '--compared-image', '-c', type=str, required=True, help='compared image')
args = parser.parse_args()


def _FSpecialGauss(size, sigma):
    """Function to mimic the 'fspecial' gaussian MATLAB function."""
    radius = size // 2
    offset = 0.0
    start, stop = -radius, radius + 1
    if size % 2 == 0:
        offset = 0.5
        stop -= 1
    x, y = np.mgrid[offset + start:stop, offset + start:stop]
    assert len(x) == size
    g = np.exp(-((x**2 + y**2) / (2.0 * sigma**2)))
    return g / g.sum()


def _SSIMForMultiScale(img1,
                       img2,
                       max_val=255,
                       filter_size=11,
                       filter_sigma=1.5,
                       k1=0.01,
                       k2=0.03):
    """Return the Structural Similarity Map between `img1` and `img2`.
  This function attempts to match the functionality of ssim_index_new.m by
  Zhou Wang: http://www.cns.nyu.edu/~lcv/ssim/msssim.zip
  Arguments:
    img1: Numpy array holding the first RGB image batch.
    img2: Numpy array holding the second RGB image batch.
    max_val: the dynamic range of the images (i.e., the difference between the
      maximum the and minimum allowed values).
    filter_size: Size of blur kernel to use (will be reduced for small images).
    filter_sigma: Standard deviation for Gaussian blur kernel (will be reduced
      for small images).
    k1: Constant used to maintain stability in the SSIM calculation (0.01 in
      the original paper).
    k2: Constant used to maintain stability in the SSIM calculation (0.03 in
      the original paper).
  Returns:
    Pair containing the mean SSIM and contrast sensitivity between `img1` and
    `img2`.
  Raises:
    RuntimeError: If input images don't have the same shape or don't have four
      dimensions: [batch_size, height, width, depth].
  """
    if img1.shape != img2.shape:
        raise RuntimeError(
            'Input images must have the same shape (%s vs. %s).', img1.shape,
            img2.shape)
    if img1.ndim != 4:
        raise RuntimeError('Input images must have four dimensions, not %d',
                           img1.ndim)

    img1 = img1.astype(np.float64)
    img2 = img2.astype(np.float64)
    _, height, width, _ = img1.shape

    # Filter size can't be larger than height or width of images.
    size = min(filter_size, height, width)

    # Scale down sigma if a smaller filter size is used.
    sigma = size * filter_sigma / filter_size if filter_size else 0

    if filter_size:
        window = np.reshape(_FSpecialGauss(size, sigma), (1, size, size, 1))
        mu1 = signal.fftconvolve(img1, window, mode='valid')
        mu2 = signal.fftconvolve(img2, window, mode='valid')
        sigma11 = signal.fftconvolve(img1 * img1, window, mode='valid')
        sigma22 = signal.fftconvolve(img2 * img2, window, mode='valid')
        sigma12 = signal.fftconvolve(img1 * img2, window, mode='valid')
    else:
        # Empty blur kernel so no need to convolve.
        mu1, mu2 = img1, img2
        sigma11 = img1 * img1
        sigma22 = img2 * img2
        sigma12 = img1 * img2

    mu11 = mu1 * mu1
    mu22 = mu2 * mu2
    mu12 = mu1 * mu2
    sigma11 -= mu11
    sigma22 -= mu22
    sigma12 -= mu12

    # Calculate intermediate values used by both ssim and cs_map.
    c1 = (k1 * max_val)**2
    c2 = (k2 * max_val)**2
    v1 = 2.0 * sigma12 + c2
    v2 = sigma11 + sigma22 + c2
    ssim = np.mean((((2.0 * mu12 + c1) * v1) / ((mu11 + mu22 + c1) * v2)))
    cs = np.mean(v1 / v2)
    return ssim, cs


def MultiScaleSSIM(img1,
                   img2,
                   max_val=255,
                   filter_size=11,
                   filter_sigma=1.5,
                   k1=0.01,
                   k2=0.03,
                   weights=None):
    """Return the MS-SSIM score between `img1` and `img2`.
  This function implements Multi-Scale Structural Similarity (MS-SSIM) Image
  Quality Assessment according to Zhou Wang's paper, "Multi-scale structural
  similarity for image quality assessment" (2003).
  Link: https://ece.uwaterloo.ca/~z70wang/publications/msssim.pdf
  Author's MATLAB implementation:
  http://www.cns.nyu.edu/~lcv/ssim/msssim.zip
  Arguments:
    img1: Numpy array holding the first RGB image batch.
    img2: Numpy array holding the second RGB image batch.
    max_val: the dynamic range of the images (i.e., the difference between the
      maximum the and minimum allowed values).
    filter_size: Size of blur kernel to use (will be reduced for small images).
    filter_sigma: Standard deviation for Gaussian blur kernel (will be reduced
      for small images).
    k1: Constant used to maintain stability in the SSIM calculation (0.01 in
      the original paper).
    k2: Constant used to maintain stability in the SSIM calculation (0.03 in
      the original paper).
    weights: List of weights for each level; if none, use five levels and the
      weights from the original paper.
  Returns:
    MS-SSIM score between `img1` and `img2`.
  Raises:
    RuntimeError: If input images don't have the same shape or don't have four
      dimensions: [batch_size, height, width, depth].
  """
    if img1.shape != img2.shape:
        raise RuntimeError(
            'Input images must have the same shape (%s vs. %s).', img1.shape,
            img2.shape)
    if img1.ndim != 4:
        raise RuntimeError('Input images must have four dimensions, not %d',
                           img1.ndim)

    # Note: default weights don't sum to 1.0 but do match the paper / matlab code.
    weights = np.array(weights if weights else
                       [0.0448, 0.2856, 0.3001, 0.2363, 0.1333])
    levels = weights.size
    downsample_filter = np.ones((1, 2, 2, 1)) / 4.0
    im1, im2 = [x.astype(np.float64) for x in [img1, img2]]
    mssim = np.array([])
    mcs = np.array([])
    for _ in range(levels):
        ssim, cs = _SSIMForMultiScale(
            im1,
            im2,
            max_val=max_val,
            filter_size=filter_size,
            filter_sigma=filter_sigma,
            k1=k1,
            k2=k2)
        mssim = np.append(mssim, ssim)
        mcs = np.append(mcs, cs)
        filtered = [
            convolve(im, downsample_filter, mode='reflect')
            for im in [im1, im2]
        ]
        im1, im2 = [x[:, ::2, ::2, :] for x in filtered]
    return (np.prod(mcs[0:levels - 1]**weights[0:levels - 1]) *
            (mssim[levels - 1]**weights[levels - 1]))


def msssim(original, compared):
    if isinstance(original, str):
        original = np.array(Image.open(original).convert('RGB'), dtype=np.float32)
    if isinstance(compared, str):
        compared = np.array(Image.open(compared).convert('RGB'), dtype=np.float32)

    original = original[None, ...] if original.ndim == 3 else original
    compared = compared[None, ...] if compared.ndim == 3 else compared

    return MultiScaleSSIM(original, compared, max_val=255)


def psnr(original, compared):
    if isinstance(original, str):
        original = np.array(Image.open(original).convert('RGB'), dtype=np.float32)
    if isinstance(compared, str):
        compared = np.array(Image.open(compared).convert('RGB'), dtype=np.float32)

    mse = np.mean(np.square(original - compared))
    psnr = np.clip(
        np.multiply(np.log10(255. * 255. / mse[mse > 0.]), 10.), 0., 99.99)[0]
    return psnr


def main():
    if args.metric != 'psnr':
        print(msssim(args.original_image, args.compared_image), end='')
    if args.metric != 'ssim':
        print(psnr(args.original_image, args.compared_image), end='')


if __name__ == '__main__':
    main()
```

Traceback
```
Traceback (most recent call last):
  File "metric.py", line 219, in <module>
    main()
  File "metric.py", line 215, in main
    print(psnr(args.original_image, args.compared_image), end='')
  File "metric.py", line 203, in psnr
    compared = np.array(Image.open(compared).convert('RGB'), dtype=np.float32)
  File "/home/khawar/anaconda3/envs/RNN_Fresh/lib/python3.5/site-packages/PIL/Image.py", line 902, in convert
    self.load()
  File "/home/khawar/anaconda3/envs/RNN_Fresh/lib/python3.5/site-packages/PIL/Jpeg2KImagePlugin.py", line 232, in load
    return ImageFile.ImageFile.load(self)
  File "/home/khawar/anaconda3/envs/RNN_Fresh/lib/python3.5/site-packages/PIL/ImageFile.py", line 280, in load
    raise_oserror(err_code)
  File "/home/khawar/anaconda3/envs/RNN_Fresh/lib/python3.5/site-packages/PIL/ImageFile.py", line 65, in raise_oserror
    raise OSError(message + " when reading image file")
OSError: broken data stream when reading image file
```

### PR DESCRIPTION
**Pull request by nulano**
Fixes #4991 
Fixes #3946

Adds support decoding subsampled RGB and YCbCr JPEG2000 images.
I left the grayscale formats as-is as I was a bit confused by the code and did not find test files for these formats.

Test images are from https://github.com/uclouvain/openjpeg-data, but these are huge and have an unusual license, so they are in https://github.com/python-pillow/pillow-depends/pull/35. `subsampling_1.jp2` and `zoo1.jp2` are YCbCr 4:2:0 and `subsampling_2.jp2` and `zoo2.jp2` are something like RGB 2:2:0 (2x2 subsampling for all 3 channels). This list is based on [this OpenJPEG test](https://github.com/uclouvain/openjpeg/blob/491299eb073c80602ca8ee333303a31a581d3fa8/tests/conformance/CMakeLists.txt#L568-L571).

~~Until the pillow-depends PR is merged, see successful test in another branch: https://github.com/nulano/Pillow/runs/1279251514?check_suite_focus=true#step:10:425~~

All four images have `color_space = OPJ_CLRSPC_UNSPECIFIED`, and based on https://github.com/uclouvain/openjpeg/issues/570 I think this may be a bug, but I did not look into fixing this.

### CODE REVIEW

**Comment by hugovk** [comment]
https://github.com/python-pillow/pillow-depends/pull/35 merged, CI restarted.

**Comment by hackermd** [comment]
@nulano I tested the `nulano:jp2-decode-subsample` branch on the image provided in #5165. It works. The image is decoded successfully and the colors are correct. However, I encountered a few compiler warnings when building from source. Will add a brief review to this PR.

**Comment by hackermd** [comment]
@nulano I am getting the following compiler warnings with clang (version 12) on Mac OSX:

```
src/_imagingft.c:1087:28: warning: assigning to 'unsigned char *' from 'char *' converts between
      pointers to integer types with different sign [-Wpointer-sign]
                    target = im->image[yy] + xx * 4;
                           ^ ~~~~~~~~~~~~~~~~~~~~~~

src/_imaging.c:3928:26: warning: comparison of integers of different signs: 'int' and 'unsigned long'
      [-Wsign-compare]
    else if ( blocks_max > SIZE_MAX/sizeof(ImagingDefaultArena.blocks_pool[0])) {
              ~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/libImaging/Resample.c:211:17: warning: comparison of integers of different signs: 'int' and
      'unsigned long' [-Wsign-compare]
    if (outSize > INT_MAX / (ksize * sizeof(double))) {
        ~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/libImaging/Draw.c:1633:31: warning: comparison of integers of different signs: 'int' and
      'unsigned long' [-Wsign-compare]
            if (outline->size > INT_MAX / sizeof(Edge)) {
                ~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~

src/libImaging/GetBBox.c:201:25: warning: assigning to 'unsigned char *' from 'char *' converts between
      pointers to integer types with different sign [-Wpointer-sign]
                  pixel = im->image[y] + x * sizeof(v);
                        ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/libImaging/RankFilter.c:75:14: warning: comparison of integers of different signs: 'int' and
      'unsigned long' [-Wsign-compare]
        size > INT_MAX / (size * sizeof(FLOAT32))) {
        ~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/libImaging/Jpeg2KDecode.c:777:41: warning: comparison of integers of different signs: 'unsigned int'
      and 'int' [-Wsign-compare]
            || tile_info.x1 - image->x0 > im->xsize
               ~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~

src/libImaging/Jpeg2KDecode.c:778:41: warning: comparison of integers of different signs: 'unsigned int'
      and 'int' [-Wsign-compare]
            || tile_info.y1 - image->y0 > im->ysize) {
               ~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~
```

```none
$ /usr/bin/cc --version
Apple clang version 12.0.0 (clang-1200.0.32.2)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

When I try to compile it with clang (version 11) on Linux (Ubuntu 20.04), compilation fails:

```none
docker build -t pillow:test . -f-<<EOF
FROM ubuntu:focal AS base

ENV LC_ALL=C.UTF-8 \
    LANG=C.UTF-8 \
    DEBIAN_FRONTEND=noninteractive \
    DEBCONF_NONINTERACTIVE_SEEN=true

RUN apt-get update && \
    apt-get install -y --no-install-suggests --no-install-recommends \
        build-essential \
        ca-certificates \
        clang-11 \
        cmake \
        curl \
        dumb-init \
        git \
        liblcms2-dev \
        libfreetype6-dev \
        libfribidi-dev \
        libharfbuzz-dev \
        libjpeg-turbo8-dev \
        libopenjp2-7-dev \
        libpng-dev \
        libtiff-dev \
        libwebp-dev \
        libxcb1-dev \
        libxcb-icccm4 \
        libxcb-image0 \
        libxcb-keysyms1 \
        libxcb-render-util0 \
        libxkbcommon-x11-0 \
        netpbm \
        libzstd-dev \
        lldb-11 \
        lld-11 \
        python3-dev \
        python3-pip \
        python3-setuptools \
        python3-tk \
        s3cmd \
        tcl8.6-dev \
        tk8.6-dev \
        unzip \
        zlib1g-dev && \
    apt-get clean && rm -rf /var/lib/apt/lists/*

WORKDIR /usr/local/share/

RUN git clone https://github.com/nulano/Pillow && \
    cd Pillow && \
    git checkout 2341c6b93306088652787335ef94ebbae6aba348 && \
    python3 -m pip install -U pip setuptools && \
    CC=clang python3 setup.py build_ext --debug install

ENTRYPOINT ["/usr/bin/dumb-init", "--", "python3"]
EOF
```

```none
$ /usr/bin/cc --version
Ubuntu clang version 11.0.0-2~ubuntu20.04.1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by nulano** [comment]
@hackermd Thank you for testing. All of those warnings are unrelated to this PR and should now be fixed by the merge @radarhere did just above your comment. You didn't provide a log for the Linux Docker build, but I'm guessing it uses `-Wall` and is failing due to the (now resolved) warnings above. If you still get warnings or compile errors after this merge, I would suggest opening a new issue.

**Comment by hackermd** [comment]
@nulano the build is already based on 2341c6b93306088652787335ef94ebbae6aba348

Here is the relevant part of the docker build log:

```none
#7 29.25 running build_ext
#7 29.29 Package libtiff-5 was not found in the pkg-config search path.
#7 29.29 Perhaps you should add the directory containing `libtiff-5.pc'
#7 29.29 to the PKG_CONFIG_PATH environment variable
#7 29.29 No package 'libtiff-5' found
#7 29.34 Package libimagequant was not found in the pkg-config search path.
#7 29.34 Perhaps you should add the directory containing `libimagequant.pc'
#7 29.34 to the PKG_CONFIG_PATH environment variable
#7 29.34 No package 'libimagequant' found
#7 29.36 Looking for `libjpeg` using pkg-config.
#7 29.36 Appending path /usr/local/share/Pillow
#7 29.36 Looking for `libopenjp2` using pkg-config.
#7 29.36 Appending path /usr/local/share/Pillow
#7 29.36 Appending path /usr/include/openjpeg-2.3
#7 29.36 Looking for `libtiff-5` using pkg-config.
#7 29.36 Looking for `libtiff-4` using pkg-config.
#7 29.36 Appending path /usr/include/x86_64-linux-gnu
#7 29.36 Looking for `zlib` using pkg-config.
#7 29.36 Looking for `freetype2` using pkg-config.
#7 29.36 Looking for `lcms2` using pkg-config.
#7 29.36 Looking for `libimagequant` using pkg-config.
#7 29.36 Appending path /usr/lib
#7 29.36 Appending path /usr/include
#7 29.36 Appending path /usr/lib/x86_64-linux-gnu
#7 29.36 Appending path /usr/local/lib
#7 29.36 Appending path /usr/local/include
#7 29.36 Looking for zlib
#7 29.36 Checking for include file zlib.h in /usr/local/share/Pillow
#7 29.36 Checking for include file zlib.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file zlib.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file zlib.h in /usr/include
#7 29.36 Found zlib.h
#7 29.36 Found library z at /usr/lib/x86_64-linux-gnu/libz.so
#7 29.36 Looking for jpeg
#7 29.36 Checking for include file jpeglib.h in /usr/local/share/Pillow
#7 29.36 Checking for include file jpeglib.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file jpeglib.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file jpeglib.h in /usr/include
#7 29.36 Found jpeglib.h
#7 29.36 Found library jpeg at /usr/lib/x86_64-linux-gnu/libjpeg.so
#7 29.36 Looking for jpeg2000
#7 29.36 Checking for openjpeg-#.# in /usr/local/share/Pillow
#7 29.36 Checking for openjpeg-#.# in /usr/include/openjpeg-2.3
#7 29.36 Checking for openjpeg-#.# in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for openjpeg-#.# in /usr/include
#7 29.36 Found openjpeg.h in /usr/include/openjpeg-2.1
#7 29.36 Best openjpeg version (2, 1) so far in /usr/include/openjpeg-2.1
#7 29.36 Found openjpeg.h in /usr/include/openjpeg-2.3
#7 29.36 Best openjpeg version (2, 3) so far in /usr/include/openjpeg-2.3
#7 29.36 Checking for openjpeg-#.# in /usr/local/include
#7 29.36 Checking for openjpeg-#.# in /usr/include/python3.8
#7 29.36 Found library openjp2 at /usr/lib/x86_64-linux-gnu/libopenjp2.so
#7 29.36 Looking for imagequant
#7 29.36 Checking for include file libimagequant.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file libimagequant.h in /usr/local/share/Pillow
#7 29.36 Checking for include file libimagequant.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file libimagequant.h in /usr/include
#7 29.36 Checking for include file libimagequant.h in /usr/local/include
#7 29.36 Checking for include file libimagequant.h in /usr/include/python3.8
#7 29.36 Looking for tiff
#7 29.36 Checking for include file tiff.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file tiff.h in /usr/local/share/Pillow
#7 29.36 Checking for include file tiff.h in /usr/include/x86_64-linux-gnu
#7 29.36 Found tiff.h
#7 29.36 Found library tiff at /usr/lib/x86_64-linux-gnu/libtiff.so
#7 29.36 Looking for freetype
#7 29.36 Found library freetype at /usr/lib/x86_64-linux-gnu/libfreetype.so
#7 29.36 Checking for include file ft2build.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file ft2build.h in /usr/include/openjpeg-2.3/freetype2
#7 29.36 Checking for include file ft2build.h in /usr/local/share/Pillow
#7 29.36 Checking for include file ft2build.h in /usr/local/share/Pillow/freetype2
#7 29.36 Checking for include file ft2build.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file ft2build.h in /usr/include/x86_64-linux-gnu/freetype2
#7 29.36 Checking for include file ft2build.h in /usr/include
#7 29.36 Checking for include file ft2build.h in /usr/include/freetype2
#7 29.36 Found ft2build.h in /usr/include/freetype2
#7 29.36 Inserting path /usr/include/freetype2
#7 29.36 Looking for lcms
#7 29.36 Checking for include file lcms2.h in /usr/include/freetype2
#7 29.36 Checking for include file lcms2.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file lcms2.h in /usr/local/share/Pillow
#7 29.36 Checking for include file lcms2.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file lcms2.h in /usr/include
#7 29.36 Found lcms2.h
#7 29.36 Found library lcms2 at /usr/lib/x86_64-linux-gnu/liblcms2.so
#7 29.36 Looking for webp
#7 29.36 Checking for include file webp/encode.h in /usr/include/freetype2
#7 29.36 Checking for include file webp/encode.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file webp/encode.h in /usr/local/share/Pillow
#7 29.36 Checking for include file webp/encode.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file webp/encode.h in /usr/include
#7 29.36 Found webp/encode.h
#7 29.36 Checking for include file webp/decode.h in /usr/include/freetype2
#7 29.36 Checking for include file webp/decode.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file webp/decode.h in /usr/local/share/Pillow
#7 29.36 Checking for include file webp/decode.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file webp/decode.h in /usr/include
#7 29.36 Found webp/decode.h
#7 29.36 Found library webp at /usr/lib/x86_64-linux-gnu/libwebp.so
#7 29.36 Looking for webpmux
#7 29.36 Checking for include file webp/mux.h in /usr/include/freetype2
#7 29.36 Checking for include file webp/mux.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file webp/mux.h in /usr/local/share/Pillow
#7 29.36 Checking for include file webp/mux.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file webp/mux.h in /usr/include
#7 29.36 Found webp/mux.h
#7 29.36 Checking for include file webp/demux.h in /usr/include/freetype2
#7 29.36 Checking for include file webp/demux.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file webp/demux.h in /usr/local/share/Pillow
#7 29.36 Checking for include file webp/demux.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file webp/demux.h in /usr/include
#7 29.36 Found webp/demux.h
#7 29.36 Found library webpmux at /usr/lib/x86_64-linux-gnu/libwebpmux.so
#7 29.36 Found library webpdemux at /usr/lib/x86_64-linux-gnu/libwebpdemux.so
#7 29.36 Couldn't find library libwebpmux in ['/usr/local/share/Pillow', '/usr/lib', '/usr/lib/x86_64-linux-gnu', '/usr/local/lib']
#7 29.36 Looking for xcb
#7 29.36 Checking for include file xcb/xcb.h in /usr/include/freetype2
#7 29.36 Checking for include file xcb/xcb.h in /usr/include/openjpeg-2.3
#7 29.36 Checking for include file xcb/xcb.h in /usr/local/share/Pillow
#7 29.36 Checking for include file xcb/xcb.h in /usr/include/x86_64-linux-gnu
#7 29.36 Checking for include file xcb/xcb.h in /usr/include
#7 29.36 Found xcb/xcb.h
#7 29.36 Found library xcb at /usr/lib/x86_64-linux-gnu/libxcb.so
#7 29.36 building 'PIL._imaging' extension
#7 29.36 creating build
#7 29.37 building 'PIL._imagingcms' extension
#7 29.37 building 'PIL._imagingft' extension
#7 29.37 creating build/temp.linux-x86_64-3.8
#7 29.37 creating build/temp.linux-x86_64-3.8
#7 29.37 creating build/temp.linux-x86_64-3.8
#7 29.37 creating build/temp.linux-x86_64-3.8/src
#7 29.37 clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -g -I/usr/include/freetype2 -I/usr/include/openjpeg-2.3 -I/usr/local/share/Pillow -I/usr/include/x86_64-linux-gnu -I/usr/include -I/usr/local/include -I/usr/include/python3.8 -c src/_imagingcms.c -o build/temp.linux-x86_64-3.8/src/_imagingcms.o
#7 29.37 building 'PIL._webp' extension
#7 29.37 creating build/temp.linux-x86_64-3.8/src
#7 29.38 creating build/temp.linux-x86_64-3.8/src
#7 29.38 clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -g -I/usr/include/freetype2 -I/usr/include/openjpeg-2.3 -I/usr/local/share/Pillow -I/usr/include/x86_64-linux-gnu -I/usr/include -I/usr/local/include -I/usr/include/python3.8 -c src/_imagingft.c -o build/temp.linux-x86_64-3.8/src/_imagingft.o
#7 29.38 creating build/temp.linux-x86_64-3.8/src/libImaging
#7 29.38 clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -g -DHAVE_WEBPMUX -I/usr/include/freetype2 -I/usr/include/openjpeg-2.3 -I/usr/local/share/Pillow -I/usr/include/x86_64-linux-gnu -I/usr/include -I/usr/local/include -I/usr/include/python3.8 -c src/_webp.c -o build/temp.linux-x86_64-3.8/src/_webp.o
#7 29.38 building 'PIL._imagingtk' extension
#7 29.39 building 'PIL._imagingmath' extension
#7 29.39 clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -g -I/usr/include/freetype2 -I/usr/include/openjpeg-2.3 -I/usr/local/share/Pillow -I/usr/include/x86_64-linux-gnu -I/usr/include -I/usr/local/include -I/usr/include/python3.8 -c src/_imagingmath.c -o build/temp.linux-x86_64-3.8/src/_imagingmath.o
#7 29.39 creating build/temp.linux-x86_64-3.8/src/Tk
#7 29.39 clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -g -I/usr/include/freetype2 -I/usr/include/openjpeg-2.3 -I/usr/local/share/Pillow -I/usr/include/x86_64-linux-gnu -I/usr/include -I/usr/local/include -I/usr/include/python3.8 -c src/_imagingtk.c -o build/temp.linux-x86_64-3.8/src/_imagingtk.o
#7 29.39 clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -g -DHAVE_LIBJPEG -DHAVE_OPENJPEG -DHAVE_LIBZ -DHAVE_LIBTIFF -DHAVE_XCB -DPILLOW_VERSION="8.1.0.dev0" -I/usr/include/freetype2 -I/usr/include/openjpeg-2.3 -I/usr/local/share/Pillow -I/usr/include/x86_64-linux-gnu -I/usr/include -I/usr/local/include -I/usr/include/python3.8 -c src/_imaging.c -o build/temp.linux-x86_64-3.8/src/_imaging.o
#7 29.40 building 'PIL._imagingmorph' extension
#7 29.40 clang -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -g -I/usr/include/freetype2 -I/usr/include/openjpeg-2.3 -I/usr/local/share/Pillow -I/usr/include/x86_64-linux-gnu -I/usr/include -I/usr/local/include -I/usr/include/python3.8 -c src/_imagingmorph.c -o build/temp.linux-x86_64-3.8/src/_imagingmorph.o
#7 29.41 error: command 'clang' failed with exit status 1
```

**Comment by nulano** [comment]
I don't see an error in that output, I think there might still be something missing. I also don't see why this PR should cause any compile issues, have you tried compiling the master branch?

**Comment by hackermd** [comment]
> I don't see an error in that output, I think there might still be something missing. I also don't see why this PR should cause any compile issues, have you tried compiling the master branch?

I am getting the same issue when trying to compile the master branch. Do I need to configure anything else for using clang to compile PIL on Ubuntu? I am fine with using GCC.

**Comment by nulano** [comment]
Is there anything blocking this?

**Comment by wiredfool** [comment]
I don't think so. There are a few uncovered lines, but we'd need to find specific j2k files to hit them.

**Comment by wiredfool** [comment]
This does interact with an upcoming security patch. It's a pretty trivial merge fix, but it changes the error from the test cases.

**Comment by wiredfool** [comment]
I think I'd prefer to merge this just after the release, so that we don't potentially trip up a security issue.

File: src/libImaging/Jpeg2KDecode.c [hunk 1]
```diff
@@ -742,19 +756,27 @@ j2k_decode_entry(Imaging im, ImagingCodecState state) {
         switch (image->numcomps) {
             case 1:
             case 2:
-                color_space = OPJ_CLRSPC_GRAY;
-                break;
+                color_space = OPJ_CLRSPC_GRAY; break;
```
Comment (radarhere) [inline src/libImaging/Jpeg2KDecode.c:759]:
```suggestion
                color_space = OPJ_CLRSPC_GRAY;
                break;
```

File: src/libImaging/Jpeg2KDecode.c [hunk 2]
```diff
@@ -742,19 +756,27 @@ j2k_decode_entry(Imaging im, ImagingCodecState state) {
...
-                color_space = OPJ_CLRSPC_SRGB;
-                break;
+                switch (subsampling) {
+                    case -1:
+                    case 0:
+                    case 3:
+                        color_space = OPJ_CLRSPC_SRGB; break;
+                    case 1:
+                    case 2:
+                        color_space = OPJ_CLRSPC_SYCC;  break;
```
Comment (radarhere) [inline src/libImaging/Jpeg2KDecode.c:769]:
```suggestion
                        color_space = OPJ_CLRSPC_SRGB;
                        break;
                    case 1:
                    case 2:
                        color_space = OPJ_CLRSPC_SYCC;
                        break;
```

**Comment by wiredfool** [comment]
I was ok with this before the release, except for the potential conflict with the security patch. Lets get it merged and fix things up if necessary.

**Review by wiredfool — APPROVED** [review]

**Pull request merged**